### PR TITLE
Supercharge Buck incremental build for Swift modules.

### PR DESCRIPTION
**Pull request by xianwen**
## TL;DR;
Make SwiftCompile implement SupportsInputBasedRuleKey properly, to avoid unnecessary Swift recompile.

Verified in Buck Sample:
https://github.com/airbnb/BuckSample/pull/132

## How Buck calculates Rule Key
When Buck see a Build Target, it invoke `buildOrFetchFromCache` to build it or fetch the build result from cache.

https://github.com/airbnb/buck/blob/dabb3ca27da483fc3f756da87a789b7d542f96f7/src/com/facebook/buck/core/build/engine/impl/CachingBuildRuleBuilder.java#L973 

### 1. It first check the hash of everything that can directly or indirectly impact the output of this build target.

This step is done via `checkMatchingLocalKey`, which returns the build result from cache if the "Default Key", a Rule Key calculated by hashing everything that can directly or indirectly impact the output of this build target, remains unchanged.

https://github.com/airbnb/buck/blob/dabb3ca27da483fc3f756da87a789b7d542f96f7/src/com/facebook/buck/core/build/engine/impl/CachingBuildRuleBuilder.java#L981

This Default Key used here is calculated by "DefaultRuleKeyFactory"
https://github.com/airbnb/buck/blob/dabb3ca27da483fc3f756da87a789b7d542f96f7/src/com/facebook/buck/rules/keys/DefaultRuleKeyFactory.java#L117

**If a cache hit is found, the build result is retrieved from the cache and returned immediately.**

### 2. It then checks the hash of all the inputs of this build target, after all its dependencies had been built.

https://github.com/airbnb/buck/blob/dabb3ca27da483fc3f756da87a789b7d542f96f7/src/com/facebook/buck/core/build/engine/impl/CachingBuildRuleBuilder.java#L1038

However, this only applies for BuildRule classes that comply with `SupportsInputBasedRuleKey` interface:

https://github.com/airbnb/buck/blob/dabb3ca27da483fc3f756da87a789b7d542f96f7/src/com/facebook/buck/core/rules/attr/SupportsInputBasedRuleKey.java#L41

This time, the Rule Key is calculated by `InputBasedRuleKeyFactory`: https://github.com/airbnb/buck/blob/dabb3ca27da483fc3f756da87a789b7d542f96f7/src/com/facebook/buck/rules/keys/InputBasedRuleKeyFactory.java#L95

**This "InputBasedRuleKey" is calculated in a different way from the "DefaultRuleKey", that only the direct inputs of the BuildRule itself is considered.**

### How the Rule Key is calculated

Rule Key is calculated by going through every field in the "Build Rule" class with `AddToRuleKey` attribute. 

For fields of basic types like `String`, `Bool`, the hash can be done directly. For more complicated objects, like the `swiftCompiler` of the "SwiftCompile" class, this process is done recursively: the `RuleKeyFactory` will look into the `CommandTool` class, and iterate through all its fields with `AddToRuleKey` fields, and add them to hash according to the same rule.

https://github.com/airbnb/buck/blob/dabb3ca27da483fc3f756da87a789b7d542f96f7/src/com/facebook/buck/swift/SwiftCompile.java#L78

### How does `getSwiftModuleOutputPath` affect the Rule Key calculation

The `getSwiftModuleOutputPath` method created in this PR, exposes the `.swiftmodule` files from a module's Swift dependencies via `cxxDeps` property which has `AddToRuleKey` attribute:
https://github.com/airbnb/buck/blob/dabb3ca27da483fc3f756da87a789b7d542f96f7/src/com/facebook/buck/swift/SwiftCompile.java#L110

Before, this `cxxDeps` would include the whole `apple-swift-compile` output directory of the dependency module, which contains `-Swift.h` and `.o` in addition to `.swiftmodule`. By limiting the exposure to only `.swiftmodule`, we avoid unnecessary rebuild of dependent modules:

image

In the mean time, if any change happened to the `-Swift.h` or the `.hmap` file, the dependent module can still detect these changes and trigger a rebuild. Because both output are still included in the Rule Key calculation via `cxxDeps` using a different Flavor, `apple-swift-objc-generated-header`:

image

### CODE REVIEW

File: src/com/facebook/buck/swift/SwiftCompile.java [hunk 1]
```diff
@@ -471,4 +474,9 @@ public SourcePath getObjCGeneratedHeaderPath() {
   public SourcePath getOutputPath() {
     return ExplicitBuildTargetSourcePath.of(getBuildTarget(), outputPath);
   }
+
+  public SourcePath getSwiftModuleOutputPath() {
```
Comment (dfed) [inline src/com/facebook/buck/swift/SwiftCompile.java:478]: can we add a `/***/` comment here to explain when this method should be used? Similar to the one above?

File: src/com/facebook/buck/swift/SwiftCompile.java [hunk 2]
```diff
@@ -471,4 +474,9 @@ public SourcePath getObjCGeneratedHeaderPath() {
   public SourcePath getOutputPath() {
```
Comment (dfed) [inline src/com/facebook/buck/swift/SwiftCompile.java:474]: Should we remove "Swift module metadata" from the comment above?

File: src/com/facebook/buck/swift/SwiftCompile.java [hunk 3]
```diff
@@ -62,12 +63,14 @@
...
 import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.Arrays;
 import java.util.Optional;
 import java.util.function.Function;
 
 /** A build rule which compiles one or more Swift sources into a Swift module. */
-public class SwiftCompile extends AbstractBuildRuleWithDeclaredAndExtraDeps {
+public class SwiftCompile extends AbstractBuildRuleWithDeclaredAndExtraDeps implements
+  SupportsInputBasedRuleKey {
```
Comment (dfed) [inline src/com/facebook/buck/swift/SwiftCompile.java:73]: Asking to learn: how is this protocol is used in this PR?

Comment (qyang-nj) [reply src/com/facebook/buck/swift/SwiftCompile.java:73]: same here

File: .gitignore [hunk 4]
```diff
@@ -1,7 +1,8 @@
 # Ant
 /ant-out/
 
-# IntelliJ build
+# IntelliJ
+.idea
```
Comment (qyang-nj) [inline .gitignore:5]: I think it's better to not have this change here. We'd like to keep a clean fork that can merge facebook master easily.

Comment (dfed) [reply .gitignore:5]: Good call. Maybe we try to push this change directly up to Facebook's repo?

File: src/com/facebook/buck/swift/SwiftCompile.java (hunk 2 above)
Comment (xianwen) [reply src/com/facebook/buck/swift/SwiftCompile.java:474]: No, the comment is still legit.

File: src/com/facebook/buck/swift/SwiftCompile.java (hunk 1 above)
Comment (xianwen) [reply src/com/facebook/buck/swift/SwiftCompile.java:478]: 👍

File: .gitignore (hunk 4 above)
Comment (xianwen) [reply .gitignore:5]: 👍

File: src/com/facebook/buck/swift/SwiftCompile.java (hunk 3 above)
Comment (xianwen) [reply src/com/facebook/buck/swift/SwiftCompile.java:73]: Will update PR description with more details.

**Comment by xianwen** [comment]
@qyang-nj @dfed , PR updated per review feedback, and the PR description had also been updated to provide some more insights about how Buck works on Rule Key calculation.
PTAL, thanks!

File: src/com/facebook/buck/swift/SwiftCompile.java [hunk 5]
```diff
@@ -471,4 +474,18 @@ public SourcePath getObjCGeneratedHeaderPath() {
...
   }
+
+  /**
+   * @return {@link SourcePath} to the .swiftmodule output from the compilation process. A
+   * swiftmodule file contains the public interface for a module, and is basically a binary file
+   * format equivalent to header files for a C framework or library.
+   *
+   * A swiftmodule file contains serialized ASTs (and possibly SIL), it conforms to
+   * Swift Binary Serialization Format, more details about this binary format can be found here:
+   * https://github.com/apple/swift/blob/master/docs/Serialization.rst.
```
Comment (dfed) [inline src/com/facebook/buck/swift/SwiftCompile.java:485]: let's link to the latest commit rather than `master` in case this URL changes
```suggestion
   * https://github.com/apple/swift/blob/7e6d62dae4bae4eb3737a6f76c0e51534c1bcca3/docs/Serialization.rst.
```

File: src/com/facebook/buck/apple/AppleLibraryDescription.java [hunk 6]
```diff
@@ -885,7 +885,7 @@ private static CxxHeaders createSwiftModuleHeaders(
     BuildTarget swiftCompileTarget = baseTarget.withAppendedFlavors(Type.SWIFT_COMPILE.getFlavor());
     SwiftCompile compile = (SwiftCompile) graphBuilder.requireRule(swiftCompileTarget);
 
-    return CxxHeadersDir.of(CxxPreprocessables.IncludeType.LOCAL, compile.getOutputPath());
+    return CxxHeadersDir.of(CxxPreprocessables.IncludeType.LOCAL, compile.getSwiftModuleOutputPath());
```
Comment (qyang-nj) [inline src/com/facebook/buck/apple/AppleLibraryDescription.java:888]: So.. how this line will change rulekey calculation?

Comment (xianwen) [reply src/com/facebook/buck/apple/AppleLibraryDescription.java:888]: @qyang-nj, I added a new section in the PR description to answer both of your questions.
